# Basic inference for microarray data

We have data for two strains of mice which we will refer to as strain 0 and 1. We want to know which genes are differentially expressed.  We extracted RNA from 12 randomely selected mice from each strain [CITE POOLING PAPER]. In one experiment we pooled the RNA from all individuals from each strain and then created 4 replicate samples from this pool. 

```{r}
library(Biobase)
library(genefilter)
library(devtools)
# install_github("dagdata","genomicsclass")
library(dagdata)
data(maPooling)
e <- maPooling
head(pData(e))

hflipt <- function(m) t(m[nrow(m):1,])
myimage <- function(m,...) {
  image(hflipt(m),xaxt="n",yaxt="n",...)
  }

myimage(as.matrix(pData(e)),col=c("white","black"))
individuals <- which(rowSums(pData(e)) == 1)

## remove replicates
names(individuals)
individuals <- individuals[-grep("tr",names(individuals))]

es <- e[,individuals]
myimage(as.matrix(pData(es)),col=c("white","black"))

es$group <- factor(as.numeric(grepl("b",colnames(es))))
es$group
```

Look at plots of gene expression across group:

```{r, fig.height=3, fig.width=6}
###look at 2 pre-selected samples for illustration
i=11425
j=11878
# install_github("rafalib","ririzarr")
library(rafalib)
mypar(1,2)
stripchart(split(exprs(es)[i,], es$group), vertical=TRUE, method="jitter", col=c(1,2), main="Gene 1", xlab="Group", pch=15)
stripchart(split(exprs(es)[j,], es$group), vertical=TRUE, method="jitter", col=c(1,2), main="Gene 2", xlab="Group", pch=15)
```

compute a t-test

```{r}
library(genefilter)
tt <- rowttests(exprs(es), es$group)
head(tt,1)
mean(exprs(es)[1,es$group == 0]) - mean(exprs(es)[1,es$group == 1]) 
diff(tapply(exprs(es)[1,], es$group, mean))
simple.t <- t.test(exprs(es)[1,] ~ es$group, var.equal=TRUE)
simple.t$p.value
tt$p.value[i]
tt$p.value[j]
mypar(1,1)
with(tt, plot(dm, -log10(p.value), 
              xlab="difference in means",
              main="'Volcano' plot"))
```


```{r}
library(limma)
?lmFit
design <- model.matrix(~ es$group)
design
fit <- lmFit(es, design)
names(fit)
head(coef(fit))
# we will introduce the eBayes() function
# in a later module called 'hierarchical modeling'
# but we call it now as it is standard in microarray analysis
fit <- eBayes(fit)
names(fit)
tt[1,]
fit$p.value[1,]
fit$t[1,]
plot(-1 * tt$statistic, fit$t[,2],
     xlab="rowttests", 
     ylab="eBayes t")
abline(0,1,col="red",lwd=3)
topTable(fit, coef=2, sort.by="p")
```
