# Dimension reduction and Heatmaps

## Dimension reduction

We start loading the tissue gene expression dataset:

```{r}
# library(devtools)
# install_github("dagdata","genomicsclass")
library(dagdata)
data(tissuesGeneExpression)
library(Biobase)
rownames(tab) <- tab$filename
t <- ExpressionSet(e, AnnotatedDataFrame(tab))
t$Tissue <- factor(t$Tissue)
colnames(t) <- paste0(t$Tissue, seq_len(ncol(t)))
```

As we noticed in the end of the clustering section, we weren't able to
"see" why the k-means algorithm defined a certain set of clusters
using only the first two genes.
TODO add text


```{r}
x <- t(exprs(t))
# TODO check
pc <- cmdscale(x)
# TODO fix
plot(pc$x[,1], pc$x[,2])
plot(pc$x[,1], pc$x[,2], type="n")
text(pc$x[,1], pc$x[,2], colnames(t))
library(rafalib)
mypar()
text(pc$x[,1], pc$x[,2], colnames(t), col=as.fumeric(t$Tissue))
```

## Heatmaps

```{r}
# install.packages("RColorBrewer")
library(RColorBrewer)
hmcol <- colorRampPalette(brewer.pal(9, "GnBu"))(100)
library(genefilter)
rv <- rowVars(exprs(t))
idx <- order(-rv)[1:40]
top <- exprs(t)[idx,]
heatmap(top, col=hmcol)
```

```{r}
# install.packages("gplots")
library(gplots)
cols <- palette(brewer.pal(8, "Dark2"))[t$Tissue]
cbind(colnames(t),cols)
heatmap.2(top, trace="none", ColSideColors=cols, col=hmcol)
```
